### PR TITLE
[feat]: MagiHuman DiT (transformer) port + parity tests (3/8)

### PR DESCRIPTION
**Pull request by SolitaryThinker**
## Summary

15B-param single-stream transformer for daVinci-MagiHuman:
- 40 layers, hidden=5120, head_dim=128, GQA `num_query_groups=8`
- Joint audio-visual denoising in a unified token sequence
- **No cross-attention** (single-stream architecture)

Pure addition. Dead code on the stack tip until the activation switch in 8/8.

## Changes

| File | LOC | Purpose |
|---|---:|---|
| `fastvideo/models/dits/magi_human.py` | 867 | Architecture port |
| `fastvideo/configs/models/dits/magi_human.py` | 110 | `MagiHumanVideoConfig` |
| `fastvideo/configs/models/dits/__init__.py` | +1 export | `MagiHumanVideoConfig` |
| `tests/local_tests/helpers/magi_human_upstream.py` | 334 | Upstream reference loader, shared by parity tests |
| `tests/local_tests/magi_human/test_magi_human_parity.py` | 281 | Base DiT parity (bit-exact, `diff_max=diff_mean=0.0`) |
| `tests/local_tests/magi_human/test_magi_human_distill_parity.py` | 245 | DMD-2 distill DiT parity (bit-exact) |

## Intentionally dropped during decomposition

Two debug-scratch files from #1280 are **not** carried into this PR:
- `tests/local_tests/magi_human/_debug_magi_human_block_parity.py`
- `tests/local_tests/magi_human/_debug_magi_human_weight_diff.py`

These were investigation scaffolding, not regression coverage.

## Verification

- `pre-commit run --files <changed paths>` ✓ (yapf, ruff, codespell, mypy green)
- DiT parity tests are GPU-gated and require a converted MagiHuman checkpoint; both are bit-exact on the source tree per #1280.
- Self-testing: ships both the DiT and its bit-exact parity tests.

## Stack context

Step **3 of 8** in the PR #1280 decomposition. Stacked on #1296.

```
... → #1295 → #1296 (t5gemma) → #THIS (3/8 DiT) → ... (4a stages) → ...
```

## Provenance

Source PR: #1280 (`will/magi` @ `4e160363`)

### CODE REVIEW

File: fastvideo/models/dits/magi_human.py [hunk 1]
```diff
@@ -0,0 +1,867 @@
...
+        dtype: torch.dtype = torch.bfloat16,
+    ):
+        super().__init__()
+        self.in_features = in_features
+        self.out_features = out_features
+        self.num_experts = num_experts
+        self.use_bias = bias
+        self.weight = nn.Parameter(
+            torch.empty(out_features * num_experts, in_features, dtype=dtype)
+        )
```
[the remark this answers is not in the document]
Comment (SolitaryThinker) [reply fastvideo/models/dits/magi_human.py:263]: Fixed in 345d2acb — switched both `weight` and `bias` to `torch.zeros`. Matches the sibling `MultiModalityRMSNorm` pattern and removes the silent-NaN failure mode if a future caller skips load_state_dict.

File: fastvideo/models/dits/magi_human.py [hunk 2]
```diff
@@ -0,0 +1,867 @@
...
+        out[text_mask] = self.text_embedder(
+            x[text_mask, : self.text_embedder.in_features].to(t_w.dtype)
+        ).to(out.dtype)
+        out[audio_mask] = self.audio_embedder(
+            x[audio_mask, : self.audio_embedder.in_features].to(a_w.dtype)
+        ).to(out.dtype)
+        out[video_mask] = self.video_embedder(
+            x[video_mask, : self.video_embedder.in_features].to(v_w.dtype)
+        ).to(out.dtype)
```
[the remark this answers is not in the document]
Comment (SolitaryThinker) [reply fastvideo/models/dits/magi_human.py:680]: I'd push back on this. MagiHuman is a single-stream joint denoiser where all three modalities (video + audio + text) are always present in the input batch — there is no degenerate "text-only" or "audio-only" code path. The empty-mask case the guards target never fires in practice, but the `.any()` check itself triggers a fresh GPU->CPU sync on every forward (one per modality, three per layer × 40 layers). The upstream reference doesn't have these guards either, and adding them would diverge from the port's parity contract. Leaving as-is.

File: fastvideo/models/dits/magi_human.py [hunk 3]
```diff
@@ -0,0 +1,867 @@
...
+    def _modality_cache_key(self, t: torch.Tensor) -> tuple:
+        return (t.data_ptr(), t.shape, t.dtype, t.device)
+
+    def forward(
+        self,
+        x: torch.Tensor,
+        coords_mapping: torch.Tensor,
+        modality_mapping: torch.Tensor,
+    ) -> torch.Tensor:
+        """
```
[the remark this answers is not in the document]
Comment (SolitaryThinker) [reply fastvideo/models/dits/magi_human.py:806]: I'd push back here. The `BaseDiT.forward` abstract signature is aspirational documentation in this codebase — existing first-class DiTs do not conform to it either. For example: `wanvideo.py` line 138 (`def forward(self, x, context, ...)`), `hunyuanvideo.py` line 64 (`def forward(self, x)`), `ltx2.py` line 86 (`def forward(self, sample, condition=None)`). The pattern is per-model signatures that mirror the reference, and the module's class docstring states this is intentional. Adding `**kwargs` would only be cosmetic since no caller passes `hidden_states`/`encoder_hidden_states` to a MagiHuman DiT — they call this signature directly with `(x, coords_mapping, modality_mapping)`. Leaving as-is.

File: fastvideo/models/dits/magi_human.py [hunk 4]
```diff
@@ -0,0 +1,867 @@
...
+            self._cached_modality_key = key
+        dispatcher = self._cached_dispatcher
+        video_mask = self._cached_video_mask
+        audio_mask = self._cached_audio_mask
+        text_mask = self._cached_text_mask
+        num_video_tokens = int(video_mask.sum().item())
+        if num_video_tokens:
+            num_frames = int(coords_mapping[:num_video_tokens, 0].max().item()) + 1
+        else:
+            num_frames = 0
```
[the remark this answers is not in the document]
Comment (SolitaryThinker) [reply fastvideo/models/dits/magi_human.py:830]: Fixed in 0d1822f7 — added `_cached_num_video_tokens` (modality-keyed) and `_cached_num_frames` (coords-keyed) following the same pattern as the existing dispatcher cache. The per-call `.item()` and `.max()` GPU->CPU syncs now collapse to a one-time cost at the first denoising step.
